File: Java/src/Word Ladder.java
import java.util.List;
import java.util.LinkedList;
public class Solution {
    public int ladderLength(String start, String end, Set<String> dict) {
        if (start.length() ==0 || end.length() == 0 || dict.size() == 0) return 0;
        if (start.equals(end)) return 1;
        
        LinkedList<String> queue = new LinkedList<>();
        LinkedList<Integer> depth = new LinkedList<>();
                
        queue.add(start);
        depth.add(1);
        
        Set<String> dist = new HashSet<>();
        dist.add(start);
        
        while (queue.size() !=0 ){
            String cur = queue.removeLast();
            int count = depth.removeLast();
            
            for(int i=0; i<cur.length(); i++){
                char[] wordChar = cur.toCharArray();
                for (char c ='a'; c<= 'z'; c++){
                    charArray[i] = c;
                    wordString = new String(charArray);
                    if (wordString.equals(end)){
                        return count+1;
                    }
                    else if(dict.contains(wordString) && !dist.contains(wordString)){
                        queue.add(wordString);
                        depth.add(count + 1);
                        dist.add(wordString);
                    }
                }
            }
        }
        return 0;
    }
}

File: Python/Word Ladder.py
class Solution:
    # @param start, a string
    # @param end, a string
    # @param dict, a set of string
    # @return an integer
    def ladderLength(self, start, end, dict):
        if not start or not end or not dict or not len(start) or not len(dict):
            return 0
        if start == end:
            return 1
        queue = [[start, 1]] 
        dist = set(start)
        while queue:
            word, dis = queue.pop(0)
            dis += 1
            for i in range(len(word)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    nw = word[:i] + c + word[i+1:]
                    if nw == end:
                        return dis 
                    if nw in dict and nw not in dist:
                        queue.append([nw, dis])
                        dist.add(nw)
        return 0
s = Solution();
print s.ladderLength("qa", "sq", ["si","go","se","cm","so","ph","mt","db","mb","sb","kr","ln","tm","le","av","sm","ar","ci","ca","br","ti","ba","to","ra","fa","yo","ow","sn","ya","cr","po","fe","ho","ma","re","or","rn","au","ur","rh","sr","tc","lt","lo","as","fr","nb","yb","if","pb","ge","th","pm","rb","sh","co","ga","li","ha","hz","no","bi","di","hi","qa","pi","os","uh","wm","an","me","mo","na","la","st","er","sc","ne","mn","mi","am","ex","pt","io","be","fm","ta","tb","ni","mr","pa","he","lr","sq","ye"])        
        
 